jicoco/src/main/java/org/jitsi/xmpp/mucclient/MucClient.java

```diff
@@ -853,4 +855,49 @@ public void pingFailed()
             }
         }
     }
+
+    private final PeriodicRunnable halfOpenConnectionPeriodicCheck = new HalfOpenConnectionPeriodicCheck();
+
+    /**
+     * @return the {@link PeriodicRunnable} that checks the XMPP connection state and triggers a disconnect if a
+     * half-open connection is detected.
+     */
+    public PeriodicRunnable getHalfOpenConnectionPeriodicCheck()
+    {
+        return halfOpenConnectionPeriodicCheck;
+    }
+
+    /**
+     * Periodically checks the connection state and triggers a disconnect if a half-open connection is detected.
+     */
+    class HalfOpenConnectionPeriodicCheck
+        extends PeriodicRunnable
+    {
+        public HalfOpenConnectionPeriodicCheck() {
+            super(2*1000*DEFAULT_PING_INTERVAL_SECONDS);
+        }
+
+        /**
+         * Triggers a disconnect if a half-open connection is detected.
+         */
+        @Override
+        public void run() {
+            super.run();
+
+            AbstractXMPPConnection con = xmppConnection;
+            if (con != null && con.isConnected() && con.isAuthenticated())
+            {
+                long lastStanzaReceivedMs = con.getLastStanzaReceived();
+                if (lastStanzaReceivedMs > 0)
+                {
+                    long nowMs = System.currentTimeMillis();
+                    if (nowMs - lastStanzaReceivedMs > 2 * 1000 * DEFAULT_PING_INTERVAL_SECONDS)
+                    {
+                        logger.warn("Half-open XMPP connection detected, will trigger a disconnect.");
+                        con.disconnect();
```

Comment:
When you trigger the disconnect, it reconnects again, right?

Reply:
Correct 👍